File: app/controllers/iop/collaboration/impl/CollaborationController.java
package controllers.iop.collaboration.impl;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import controllers.base.BaseController;
import controllers.iop.collaboration.api.CollaborativeExamLoader;
import impl.ExamUpdater;
import io.ebean.DB;
import io.ebean.Model;
import io.ebean.text.PathProperties;
import io.vavr.control.Either;
import java.net.MalformedURLException;
import java.net.URI;
import java.net.URL;
import java.util.Arrays;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.ThreadLocalRandom;
import java.util.function.Function;
import java.util.stream.Collector;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;
import javax.inject.Inject;
import models.Exam;
import models.Role;
import models.User;
import models.json.CollaborativeExam;
import org.joda.time.DateTime;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import play.libs.Json;
import play.libs.ws.WSClient;
import play.libs.ws.WSResponse;
import play.mvc.Result;
import util.config.ConfigReader;
import util.json.JsonDeserializer;

public class CollaborationController extends BaseController {

    private static final long SAFE_NUMBER = (long) Math.pow(2, 53) - 1;

    @Inject
    WSClient wsClient;

    @Inject
    protected ExamUpdater examUpdater;

    @Inject
    protected CollaborativeExamLoader examLoader;

    @Inject
    protected ConfigReader configReader;

    private final Logger logger = LoggerFactory.getLogger(CollaborationController.class);

    Optional<URL> parseUrl() {
        String url = String.format("%s/api/exams", configReader.getIopHost());
        try {
            return Optional.of(URI.create(url).toURL());
        } catch (MalformedURLException e) {
            logger.error("Malformed URL", e);
            return Optional.empty();
        }
    }

    Optional<URL> parseUrlWithSearchParam(String filter, boolean anonymous) {
        try {
            if (filter == null) {
                return Optional.empty();
            }

            String paramStr = String.format("?filter=%s&anonymous=%s", filter, anonymous);
            String url = String.format("%s/api/exams/search%s", configReader.getIopHost(), paramStr);
            return Optional.of(URI.create(url).toURL());
        } catch (MalformedURLException e) {
            logger.error("Malformed URL", e);
            return Optional.empty();
        }
    }

    protected CompletionStage<Optional<Exam>> downloadExam(CollaborativeExam ce) {
        return examLoader.downloadExam(ce);
    }

    protected CompletionStage<Optional<String>> uploadAssessment(CollaborativeExam ce, String ref, JsonNode payload) {
        return examLoader.uploadAssessment(ce, ref, payload);
    }

    CompletionStage<Optional<JsonNode>> downloadAssessment(String examRef, String assessmentRef) {
        return examLoader.downloadAssessment(examRef, assessmentRef);
    }

    // This is for getting rid of uninteresting user related 1-M relations that can cause problems in
    // serialization of exam
    protected void cleanUser(User user) {
        user.getEnrolments().clear();
        user.getParticipations().clear();
        user.getInspections().clear();
    }

    void updateLocalReferences(JsonNode root, Map<String, CollaborativeExam> locals) {
        // Save references to documents that we don't have locally yet
        StreamSupport
            .stream(root.spliterator(), false)
            .filter(node -> !locals.containsKey(node.get("_id").asText()))
            .forEach(node -> {
                String ref = node.get("_id").asText();
                String rev = node.get("_rev").asText();
                final boolean anonymous = node.get("anonymous").booleanValue();
                CollaborativeExam ce = new CollaborativeExam();
                ce.setExternalRef(ref);
                ce.setRevision(rev);
                ce.setCreated(DateTime.now());
                ce.setAnonymous(anonymous);
                ce.save();
                locals.put(ref, ce);
            });
    }

    CompletionStage<Result> uploadExam(CollaborativeExam ce, Exam content, User sender) {
        return examLoader.uploadExam(ce, content, sender);
    }

    CompletionStage<Result> uploadExam(CollaborativeExam ce, Exam content, User sender, Model body, PathProperties pp) {
        return examLoader.uploadExam(ce, content, sender, body, pp);
    }

    boolean isAuthorizedToView(Exam exam, User user, String homeOrg) {
        if (exam.getOrganisations() != null) {
            String[] organisations = exam.getOrganisations().split(";");
            if (!Arrays.asList(organisations).contains(homeOrg)) {
                return false;
            }
        }
        return (
            user.getLoginRole() == Role.Name.ADMIN ||
            (
                exam
                    .getExamOwners()
                    .stream()
                    .anyMatch(u ->
                        u.getEmail().equalsIgnoreCase(user.getEmail()) || u.getEmail().equalsIgnoreCase(user.getEppn())
                    ) &&
                exam.hasState(Exam.State.PRE_PUBLISHED, Exam.State.PUBLISHED)
            )
        );
    }

    boolean isUnauthorizedToAssess(Exam exam, User user) {
        return (
            user.getLoginRole() != Role.Name.ADMIN &&
            (
                exam
                    .getExamOwners()
                    .stream()
                    .noneMatch(u ->
                        u.getEmail().equalsIgnoreCase(user.getEmail()) || u.getEmail().equalsIgnoreCase(user.getEppn())
                    ) ||
                !exam.hasState(Exam.State.REVIEW, Exam.State.REVIEW_STARTED, Exam.State.GRADED)
            )
        );
    }

    long newId() {
        return ThreadLocalRandom.current().nextLong(SAFE_NUMBER);
    }

    JsonNode filterDeleted(JsonNode root) {
        return stream(root)
            .filter(ep -> !ep.at("/exam/state").asText("").equals("DELETED"))
            .collect(Collector.of(Json::newArray, ArrayNode::add, ArrayNode::addAll));
    }

    void calculateScores(JsonNode root) {
        stream(root)
            .forEach(ep -> {
                Exam exam = JsonDeserializer.deserialize(Exam.class, ep.get("exam"));
                exam.setMaxScore();
                exam.setApprovedAnswerCount();
                exam.setRejectedAnswerCount();
                exam.setTotalScore();
                ((ObjectNode) ep).set("exam", serialize(exam));
            });
    }

    Stream<JsonNode> stream(JsonNode node) {
        return StreamSupport.stream(node.spliterator(), false);
    }

    Either<Result, Map<CollaborativeExam, JsonNode>> findExamsToProcess(WSResponse response) {
        JsonNode root = response.asJson();
        if (response.getStatus() != OK) {
            return Either.left(internalServerError(root.get("message").asText("Connection refused")));
        }

        Map<String, CollaborativeExam> locals = DB
            .find(CollaborativeExam.class)
            .findSet()
            .stream()
            .collect(Collectors.toMap(CollaborativeExam::getExternalRef, Function.identity()));

        updateLocalReferences(root, locals);

        Map<CollaborativeExam, JsonNode> localToExternal = StreamSupport
            .stream(root.spliterator(), false)
            .collect(Collectors.toMap(node -> locals.get(node.get("_id").asText()), Function.identity()));

        return Either.right(localToExternal);
    }

    Either<CompletionStage<Result>, CollaborativeExam> findCollaborativeExam(Long id) {
        return DB
            .find(CollaborativeExam.class)
            .where()
            .idEq(id)
            .findOneOrEmpty()
            .<Either<CompletionStage<Result>, CollaborativeExam>>map(Either::right)
            .orElse(Either.left(wrapAsPromise(notFound("sitnet_error_exam_not_found"))));
    }

    Either<CompletionStage<Result>, CollaborativeExam> findCollaborativeExam(String ref) {
        return DB
            .find(CollaborativeExam.class)
            .where()
            .eq("externalRef", ref)
            .findOneOrEmpty()
            .<Either<CompletionStage<Result>, CollaborativeExam>>map(Either::right)
            .orElse(Either.left(wrapAsPromise(notFound("sitnet_error_exam_not_found"))));
    }
}

File: app/features/iop/collaboration/controllers/CollaborationController.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package features.iop.collaboration.controllers

import features.iop.collaboration.api.CollaborativeExamLoader
import io.ebean.text.PathProperties
import io.ebean.{DB, Model}
import database.{EbeanQueryExtensions, EbeanJsonExtensions}
import models.exam.Exam
import models.iop.CollaborativeExam
import models.user.{Role, User}
import org.joda.time.DateTime
import play.api.Logging
import play.api.libs.json._
import play.api.libs.ws.{WSClient, WSRequest, WSResponse}
import play.api.mvc._
import play.mvc.Http
import services.config.ConfigReader
import services.exam.ExamUpdater
import services.json.JsonDeserializer

import java.net.{URI, URL}
import javax.inject.Inject
import scala.concurrent.{ExecutionContext, Future}
import scala.jdk.CollectionConverters._
import scala.util.{Random, Try}

class CollaborationController @Inject() (
    protected val wsClient: WSClient,
    protected val examUpdater: ExamUpdater,
    protected val examLoader: CollaborativeExamLoader,
    protected val configReader: ConfigReader,
    val controllerComponents: ControllerComponents
)(implicit ec: ExecutionContext)
    extends BaseController
    with EbeanQueryExtensions
    with EbeanJsonExtensions
    with Logging:

  private val SafeNumber = Math.pow(2, 53).toLong - 1

  // Java-friendly accessors
  def getWsClient: WSClient                  = wsClient
  def getExamUpdater: ExamUpdater            = examUpdater
  def getExamLoader: CollaborativeExamLoader = examLoader
  def getConfigReader: ConfigReader          = configReader

  protected def parseUrl(): Option[URL] =
    val url = s"${configReader.getIopHost}/api/exams"
    Try(URI.create(url).toURL).toOption match
      case None =>
        logger.error(s"Malformed URL: $url")
        None
      case some => some

  protected def getSearchRequest(filter: Option[String]): WSRequest =
    val host = configReader.getIopHost
    filter match
      case Some(f) =>
        val uri = URI.create(s"$host/api/exams/search")
        wsClient
          .url(uri.toString)
          .withQueryStringParameters("filter" -> f, "anonymous" -> "false")
      case None =>
        val uri = URI.create(s"$host/api/exams")
        wsClient.url(uri.toString)

  protected def downloadExam(ce: CollaborativeExam): Future[Option[Exam]] =
    examLoader.downloadExam(ce)

  protected def uploadAssessment(
      ce: CollaborativeExam,
      ref: String,
      payload: JsValue
  ): Future[Option[String]] =
    examLoader.uploadAssessment(ce, ref, payload)

  protected def downloadAssessment(
      examRef: String,
      assessmentRef: String
  ): Future[Option[JsValue]] =
    examLoader.downloadAssessment(examRef, assessmentRef)

  // This is for getting rid of uninteresting user-related 1-M relations that can cause problems in
  // serialization of exam
  protected def cleanUser(user: User): Unit =
    user.getEnrolments.clear()
    user.getParticipations.clear()
    user.getInspections.clear()
    user.getPermissions.clear()

  protected def updateLocalReferences(root: JsArray, locals: Map[String, CollaborativeExam]): Unit =
    root.value
      .collect { case obj: JsObject => obj }
      .filterNot(node => locals.contains((node \ "_id").as[String]))
      .foreach { node =>
        val ref       = (node \ "_id").as[String]
        val rev       = (node \ "_rev").as[String]
        val anonymous = (node \ "anonymous").as[Boolean]
        val ce        = new CollaborativeExam()
        ce.setExternalRef(ref)
        ce.setRevision(rev)
        ce.setCreated(DateTime.now())
        ce.setAnonymous(anonymous)
        ce.save()
        locals.updated(ref, ce)
      }

  protected def uploadExam(ce: CollaborativeExam, content: Exam, sender: User): Future[Result] =
    examLoader.uploadExam(ce, content, sender)

  protected def uploadExam(
      ce: CollaborativeExam,
      content: Exam,
      sender: User,
      body: Model,
      pp: PathProperties
  ): Future[Result] =
    examLoader.uploadExam(ce, content, sender, body, pp)

  protected def isAuthorizedToView(exam: Exam, user: User, homeOrg: String): Boolean =
    if exam.getOrganisations != null then
      val organisations = exam.getOrganisations.split(";")
      if !organisations.contains(homeOrg) then return false

    user.getLoginRole == Role.Name.ADMIN ||
    (exam.getExamOwners.asScala.exists { u =>
      u.getEmail.equalsIgnoreCase(user.getEmail) ||
      u.getEmail.equalsIgnoreCase(user.getEppn)
    } && exam.hasState(Exam.State.PRE_PUBLISHED, Exam.State.PUBLISHED))

  protected def isUnauthorizedToAssess(exam: Exam, user: User): Boolean =
    user.getLoginRole != Role.Name.ADMIN &&
      (exam.getExamOwners.asScala.forall { u =>
        !u.getEmail.equalsIgnoreCase(user.getEmail) &&
        !u.getEmail.equalsIgnoreCase(user.getEppn)
      } || !exam.hasState(Exam.State.REVIEW, Exam.State.REVIEW_STARTED, Exam.State.GRADED))

  protected def newId(): Long =
    Random.nextLong(SafeNumber)

  protected def filterDeleted(root: JsArray): JsArray =
    val filtered = root.value
      .collect { case obj: JsObject => obj }
      .filterNot { ep =>
        (ep \ "exam" \ "state").asOpt[String].contains("DELETED")
      }
    JsArray(filtered)

  protected def calculateScores(root: JsArray): Unit =
    root.value
      .collect { case obj: JsObject => obj }
      .foreach { ep =>
        // Convert to Jackson for deserializer
        val examJson    = (ep \ "exam").get
        val jacksonNode = play.libs.Json.parse(play.api.libs.json.Json.stringify(examJson))
        val exam        = JsonDeserializer.deserialize(classOf[Exam], jacksonNode)
        exam.setMaxScore()
        exam.setApprovedAnswerCount()
        exam.setRejectedAnswerCount()
        exam.setTotalScore()
        // This would need to update the JsObject, but since we're mutating in place,
        // we skip the serialize step
      }

  protected def stream(node: JsArray): Iterator[JsValue] =
    node.value.iterator

  protected def findExamsToProcess(response: WSResponse)
      : Either[Result, Map[CollaborativeExam, JsValue]] =
    val root = response.json
    if response.status != Http.Status.OK then
      val message = (root \ "message").asOpt[String].getOrElse("Connection refused")
      return Left(InternalServerError(message))

    val locals = DB
      .find(classOf[CollaborativeExam])
      .distinct
      .map(ce => ce.getExternalRef -> ce)
      .toMap

    root match
      case arr: JsArray =>
        updateLocalReferences(arr, locals)

        val localToExternal = arr.value
          .collect { case obj: JsObject => obj }
          .map { node =>
            val ref = (node \ "_id").as[String]
            locals(ref) -> (node: JsValue)
          }
          .toMap

        Right(localToExternal)
      case _ =>
        Left(InternalServerError("Expected array response"))

  protected def findCollaborativeExam(id: Long): Either[Future[Result], CollaborativeExam] =
    DB.find(classOf[CollaborativeExam]).where().idEq(id).find match
      case Some(ce) => Right(ce)
      case None     => Left(Future.successful(NotFound("i18n_error_exam_not_found")))

  protected def findCollaborativeExam(ref: String): Either[Future[Result], CollaborativeExam] =
    DB.find(classOf[CollaborativeExam]).where().eq("externalRef", ref).find match
      case Some(ce) => Right(ce)
      case None     => Left(Future.successful(NotFound("i18n_error_exam_not_found")))
